### ISSUE TITLE
[Bug]: [react/prop-types] False positive when wrapped with React.memo

### ISSUE THREAD

**Issue by NuclleaR** [jsx-eslint/eslint-plugin-react#3866]
### Is there an existing issue for this?

- [x] I have searched the existing issues and my issue is unique
- [x] My issue appears in the command-line and not only in the text editor

### Description Overview

eslint v8.57.1
eslint-plugin-react: 7.37.2

![Image](https://github.com/user-attachments/assets/1c609b1e-cea2-436e-b8a0-01a30ca14505)

![Image](https://github.com/user-attachments/assets/06e035ce-700a-47f2-be9d-e4ba65ab7ad9)

### eslint-plugin-react version

v7.37.2

### eslint version

v8.57.1

### node version

v18.17.1

**Comment by NuclleaR** [comment]
@rjgotten this is not how it should work. If React allows us to use generics, eslint should respect them

**Comment by ljharb** [comment]
@NuclleaR sure but someone would have to manually update the rule to support the generics, which are a recent addition to the TS types. Are you volunteering?

**Comment by rjgotten** [comment]
> @NuclleaR 
> [@rjgotten](https://github.com/rjgotten) this is not how it should work. If React allows us to use generics, eslint should respect them

Oh, different from some of the maintainers - apparently, I actually agree with you there.
It's a bug. Clear as day.
It means the lint rule is poorly written and is taking shortcuts to establish the parameter's concrete type, that it shouldn't be. It was always written wrong. It just so happened to not fail while `React.memo` didn't have a generic parameter.

I was just giving you a potential workaround.

**Comment by rjgotten** [comment]
> @ljharb 
> [@rjgotten](https://github.com/rjgotten) no, it means the lint rule was written before the TS type took generics, and nobody noticed to update it until this issue.

No, it means the rule was implemented poorly.
If it was implemented correctly, it would've invoked the TypeScript type checker API, which you simply pass the relevant node (the props parameter) and the type checker would spit out the node's correct type according to the type checker. No matter whether the type would be set explicitly or would come from setting the generic constraint.

See also:
https://typescript-eslint.io/developers/custom-rules#typed-rules

Concrete relevant lines being:
```ts
const services = ESLintUtils.getParserServices(context);
const type = services.getTypeAtLocation(node);
```

If the rule were implemented correctly, it would have been implemented as _two_ rules.
A non-typed base eslint rule.
And a typed rule relying on typescript-eslint and its integration facilities with the TypeScript type checker.

**Comment by ljharb** [comment]
@rjgotten PRs are welcome, but that is not the correct description of how a rule should be implemented, because this is a JavaScript project. It is not a bug, it’s just a missing enhancement.
